### PR TITLE
Use native AAs for DsymbolTable.tab

### PR DESCRIPTION
**Pull request by ibuclaw**
Continuation of #5428.

### CODE REVIEW

**Review by wilzbach — APPROVED** [review]

**Comment by rainers** [comment]
Have you measured the impact on performance? I opted for the dmd specific ones in dmd.dmangle because they were slightly faster in my benchmarks.

**Comment by ibuclaw** [comment]
Well I'd imagine that you'd find the internal compiler AA to be a little bit faster.  Being able to inline the implementation helps greatly in that matter.

In terms of what this is doing vs. the internal AA, there won't be much to it.  `void*` is used as the key type so that we're not calling the `getHash()`/`toHash()` implementation for classes.

**Comment by rainers** [comment]
Yeah, knowing to deal with void* only helps to avoid all the indirections through the type info.

When I tried it, I was hoping for the quadratic probing of the native AAs to help, but even the compiler internal AAs (that still use linked lists per bucket) didn't benefit from a rewrite in that direction.

**Comment by ibuclaw** [comment]
Things might get more interesting anyway.  I would like to see some progression in this area also (https://github.com/dlang/dmd/pull/4569/commits/e6c838925665039a2cb98a9ff5cdc6a6313571a9)

**Review by JinShil — APPROVED** [review]

File: src/dmd/dsymbol.d [hunk 1]
```diff
@@ -2078,44 +2077,40 @@ extern (C++) final class ForwardingScopeDsymbol : ScopeDsymbol
...
     Dsymbol insert(Dsymbol s)
     {
         //printf("DsymbolTable::insert(this = %p, '%s')\n", this, s.ident.toChars());
-        const ident = s.ident;
-        Dsymbol* ps = cast(Dsymbol*)dmd_aaGet(&tab, cast(void*)ident);
-        if (*ps)
+        if (cast(void*)s.ident in tab)
             return null; // already in table
-        *ps = s;
+        tab[cast(void*)s.ident] = s;
```
Comment (WalterBright) [inline src/dmd/dsymbol.d]: The trouble here is this adds a second lookup. This is not acceptable (performance reasons).

File: src/dmd/dsymbol.d [hunk 2]
```diff
@@ -2078,44 +2077,40 @@ extern (C++) final class ForwardingScopeDsymbol : ScopeDsymbol
...
     // when ident and s are not the same
     Dsymbol insert(const Identifier ident, Dsymbol s)
     {
         //printf("DsymbolTable::insert()\n");
-        Dsymbol* ps = cast(Dsymbol*)dmd_aaGet(&tab, cast(void*)ident);
-        if (*ps)
+        if (cast(void*)ident in tab)
             return null; // already in table
-        *ps = s;
+        tab[cast(void*)ident] = s;
```
Comment (WalterBright) [inline src/dmd/dsymbol.d]: Another double lookup.

**Review by WalterBright — CHANGES_REQUESTED** [review]
Cannot accept unless the redundant lookups are fixed.

File: src/dmd/dsymbol.d (hunk 1 above)
Comment (ibuclaw) [reply src/dmd/dsymbol.d]: You were happy to add one to #5428.

Comment (WalterBright) [reply src/dmd/dsymbol.d]: Yes, and that wasn't in a hot section of code. The dmd symbol table is very hot.

Comment (ibuclaw) [reply src/dmd/dsymbol.d]: Well in the absence of a getOrCreate property for built-in AAs, not much can really be done to help that there.

Comment (rainers) [reply src/dmd/dsymbol.d]: Well, there is _aaGetY which could be called directly, but I'm not sure we should go that road. It is already used by `dup`, maybe a similar small helper in object.d would be a nice extension. IIRC there are also a couple of enhancement requests in bugzilla for this.

Comment (ibuclaw) [reply src/dmd/dsymbol.d]: Well, `_aaGetY` does not have enough information to say whether or not a value existed before a lookup was requested.  One would have to either define a new internal function, move the common bits into new helpers, and add the difference

i.e:
```
void* _aaGetY(...)
{
  auto p = aaLookup(...);
  if (!p)
    p = aaInsert(...);
  return p;
}

void* _aaGetOrInsert(..., ref bool existed)
{
  existed = false;
  auto p = aaLookup(...);
  if (!p)
    p = aaInsert(...);
  else
    existed = true;
  return p;
}
```

Comment (rainers) [reply src/dmd/dsymbol.d]: In general the information `existed` is desirable, but for dmd, _aaGetY has the same semantics as dmd_aaGet. The zero initialization is used to detect whether the key has been added.

**Comment by WalterBright** [comment]
I'm going to Phantom Zone this until a solution is found for the double lookups, at which point it can be revived. The penalty of double lookup outweighs the advantage of dogfooding.

**Pull request closed without merging**